hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java

```diff
@@ -659,34 +716,123 @@ private Predicate getEqualToPredicate(final Object transformedValue, final Path<
         }
 
         private Predicate getNotEqualToPredicate(final Object transformedValue, final Path<Object> fieldPath,
-                final Database database) {
+                final String finalProperty, final A enumField) {
+
+            if (transformedValue == null) {
+                return toNotNullPredicate(fieldPath);
+            }
+
             if (transformedValue instanceof String) {
                 if (StringUtils.isEmpty(transformedValue)) {
-                    return cb.and(cb.isNotNull(pathOfString(fieldPath)), cb.notEqual(pathOfString(fieldPath), ""));
+                    return toNotNullOrEmptyPredicate(fieldPath);
                 }
 
-                final String preFormattedValue = escapeValueToSQL((String) transformedValue, database, ESCAPE_CHAR);
-                return cb.notLike(cb.upper(pathOfString(fieldPath)), preFormattedValue.toUpperCase(), ESCAPE_CHAR);
+                final String sqlValue = toSQL((String) transformedValue);
+                final String[] fieldNames = getSubAttributesFrom(finalProperty);
+
+                if (isSimpleField(fieldNames, enumField.isMap())) {
+                    return toNullOrNotEqualPredicate(fieldPath, sqlValue);
+                }
+
+                clearOuterJoinsIfNotNeeded();
+
+                return toNotEqualWithSubQueryPredicate(enumField, sqlValue, fieldNames);
             }
 
-            if (transformedValue == null) {
-                return cb.isNotNull(pathOfString(fieldPath));
+            return toNotEqualPredicate(fieldPath, transformedValue);
+        }
+
+        private void clearOuterJoinsIfNotNeeded() {
+            if (!joinsNeeded) {
+                root.getJoins().clear();
             }
+        }
+
+        private Predicate toNotNullPredicate(final Path<Object> fieldPath) {
+            return cb.isNotNull(pathOfString(fieldPath));
+        }
 
+        private Predicate toNotEqualPredicate(final Path<Object> fieldPath, final Object transformedValue) {
             return cb.notEqual(fieldPath, transformedValue);
         }
 
-        private static String escapeValueToSQL(final String transformedValue, final Database database,
-                final char escapeChar) {
+        private Predicate toNullOrNotEqualPredicate(final Path<Object> fieldPath, final String sqlValue) {
+            return cb.or(cb.isNull(pathOfString(fieldPath)),
+                    cb.notLike(cb.upper(pathOfString(fieldPath)), sqlValue, ESCAPE_CHAR));
+        }
+
+        private Predicate toNotNullOrEmptyPredicate(final Path<Object> fieldPath) {
+            return cb.and(cb.isNotNull(pathOfString(fieldPath)), cb.notEqual(pathOfString(fieldPath), ""));
+        }
+
+        @SuppressWarnings({ "rawtypes", "unchecked" })
+        private Predicate toNotEqualWithSubQueryPredicate(final A enumField, final String sqlValue,
```

Comment:
try to refactor in order to remove a lot of duplication with _toOutWithSubQueryPredicate_ method, e.g. you could extract the common code in separate method

Reply:
Refactoring seems to be not possible here as the similar code is the creation of subquery/innerFieldPath etc. But these values are needed to create the predicates that are the different parts